### ISSUE TITLE
Picture in Picture Mode on Firefox

### ISSUE THREAD

**Issue by raksdewji** [ianramzy/decentralized-video-chat#45]
Is there any chance we can get this option included for Firefox users?

**Comment by Chaphasilor** [comment]
It's not working on Firefox for Desktop? 👀

**Comment by raksdewji** [comment]
Currently getting this error when trying to toggle PiP
![image](https://user-images.githubusercontent.com/18670250/83427440-f9198f80-a3ed-11ea-8045-da47a6abfc58.png)

**Comment by Chaphasilor** [comment]
Okay, can you try right-clicking on the other participant's video and seeing if PiP is available through the context menu?

It seems like Firefox hasn't implemented the W3C PiP API yet because it's still a draft, but I'll check later if that is actually still the case!

**Comment by raksdewji** [comment]
Tested on Linux and Windows and both had no PiP in the context menu.

![image](https://user-images.githubusercontent.com/18670250/83916314-35bcf200-a732-11ea-95c8-30381968d377.png)

**Comment by Chaphasilor** [comment]
Thanks for checking. I'll look into it tomorrow and see what options we have :)

**Comment by Chaphasilor** [comment]
Okay, so I just spent a almost two hours investigating, and I have bad news.

tl;dr: Firefox doesn't yet support picture in picture well enough so that we can use it in Zipcall :/

The issue here is twofold:  
1. Firefox is still using their own, proprietary implementation on pip. There is no API for toggling pip, only user gestures (fly-outs, context menus, etc.). That wouldn't be a huge problem of its own, but then there's the second problem:
2. Their implementation doesn't seem to support more complex video sources, like webcam inputs or WebRTC streams yet. Only 'normal' videos with sources pointing either to a static video file or an m3u8-playlist file are supported. Because of this, we are unable to enable or provide a pip-mode in Firefox for now, it simply isn't possible from what I've found.  
  There sadly isn't any official statement about that and very few info in general (aside from a StackOverflow question with negative score, no comments and no answers), so I might be wrong, but I don't think this is worth the effort right now.

Sorry to disappoint you :/

**Comment by axtux** [comment]
Firefox does not support picture in picture the way you could use it but the statement is wrong: Firefox does support picture in picture. I think the message should be updated to something like: Your browser does not support picture in picture for WebRTC streams. And maybe add a link to this issue if Firefox is detected?

**Comment by Chaphasilor** [comment]
@axtux yeah you're right, the error could be more specific. PRs are welcome ;)
